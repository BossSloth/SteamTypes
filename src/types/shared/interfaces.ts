/* eslint-disable @typescript-eslint/no-explicit-any */
export interface ObservableValue<T> {
  /**
   * Used in {@link Set} to compare the new value with the current value.
   * If it returns true, {@link Set} will not notify subscribers.
   * If this is undefined, {@link Set} will do a strict equality check (===).
   */
  m_fnEquals?(value1: T, value2: T): boolean;

  /**
   * Sets the value of the observable.
   * If the value is different from the current value, it will notify subscribers.
   */
  Set(value: T): void;

  /**
   * Subscribes to changes in the value.
   * @param callback The callback to be called when the value changes.
   * @returns An object with an `Unsubscribe` method to remove the subscription.
   */
  Subscribe(callback: (value: T) => void): { Unsubscribe: () => void; };

  m_callbacks: Callbacks<(value: T) => void>;

  m_currentValue: T;

  /**
   * Returns the number of subscribers from {@link m_callbacks.CountRegistered}.
   */
  SubscriberCount: number;

  /**
   * Returns the current value from {@link m_currentValue}.
   */
  Value: T;
}

// TODO: update converter to also detect this type. Make it default to `<unknown, getType()>`
/**
 * @exported
 */
export interface ReducedValue<TInput, TOutput> {
  /**
   * Creates a new input into the reduced value.
   * @param debugName Optional debug name for this input.
   * @returns An object with `Set` and `Delete` methods to manage this input.
   */
  CreateInput(debugName?: string): {
    Set: (value: TInput) => void;

    Delete: () => void;
  };

  /**
   * The reducer function used to combine all input values into a single output.
   */
  m_fnReducer(inputs: TInput[]): TOutput;

  /**
   * Forces recomputation of the reduced value from all current inputs.
   */
  RecalculateValue(): void;

  /**
   * The internal log instance used for debugging.
   */
  Log: Log;

  /**
   * Internal map of all input values, keyed by their assigned ID.
   */
  m_mapInputs: Map<number, {
    value: TInput;

    strDebugCallstack: string | null;

    strDebugName: string;
  }>;

  /**
   * Incrementing ID counter for assigning keys to inputs.
   */
  m_nNextInputKey: number;

  /**
   * Optional name for debugging/logging.
   */
  m_strNameForLog: string | null;

  /**
   * The underlying observable that stores the reduced value.
   */
  m_SubscribableValue: ObservableValue<TOutput>;

  /**
   * Provides access to the internal subscribable reduced value.
   */
  SubscribableValue: ObservableValue<TOutput>;

  /**
   * Returns the current reduced value.
   */
  Value: TOutput;
}

/**
 * @exported
 */
export interface MappedObservable<TSource, TMapped> {
  /** @native */
  GetValue(): unknown;

  /**
   * The mapping function applied to the source value.
   * Whenever the source changes, this is used to compute the mapped value.
   */
  m_fnMap(value: TSource): TMapped;

  /**
   * Subscribes to changes in the mapped value.
   * @param callback The callback to be called when the mapped value changes.
   * @returns An object with an `Unsubscribe` method to remove the subscription.
   */
  Subscribe(callback: (value: TMapped) => void): { Unsubscribe: () => void; };

  /** @native */
  SyncStore(): unknown;

  /**
   * Forces recomputation of the mapped value from the original source.
   * Also updates the internal {@link m_mappedSubscribableValue}.
   */
  UpdateMappedValue(): void;

  /**
   * Flag indicating whether the mapped value needs to be recomputed.
   * If true, {@link UpdateMappedValue} will be called before returning {@link Value}.
   */
  m_bMappedValueStale: boolean;

  /**
   * The internal observable that stores the mapped value and handles subscriptions.
   */
  m_mappedSubscribableValue: ObservableValue<TMapped>;

  /**
   * The original source observable that this mapped value depends on.
   */
  m_originalSubscribableValue: ObservableValue<TSource>;

  /**
   * Returns the current mapped value.
   * If the value is stale, recomputes it first.
   */
  Value: TMapped;
}

export interface Callbacks<T extends (...args: any) => unknown = () => void> {
  /**
   * Removes all registered callbacks in {@link m_vecCallbacks}.
   */
  ClearAllCallbacks(): void;

  /**
   * @returns the number of registered callbacks in {@link m_vecCallbacks}.
   */
  CountRegistered(): number;

  /**
   * Calls all registered callbacks with the given arguments.
   *
   * @param args The arguments to pass to the callbacks.
   */
  Dispatch(...args: Parameters<T>): void;

  Register(callback: T): { Unregister: () => void; };

  m_vecCallbacks: T[];
}

/**
 * @exported
 */
export interface Unsubscribable {
  /**
   * Unregisters the callback.
   */
  Unregister(): void;
}

export interface Log {
  Assert(): unknown;

  Debug(): unknown;

  Error(): unknown;

  Info(): unknown;

  IsDebugEnabled(): unknown;

  Log(e: unknown, ...t: unknown[]): void;

  m_fnIdGenerator(): unknown;

  Warning(): unknown;

  m_sName: string;
}
