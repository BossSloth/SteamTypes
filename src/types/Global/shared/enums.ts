/** @public */
export enum ESteamUISound {
  LaunchGame,
  FriendMessage,
  ChatMention,
  ChatMessage,
  ToastMessage,
  ToastAchievement,
  ToastMisc,
  ToastMiscShort,
  FriendOnline,
  FriendInGame,
  VolSound,
  ShowModal,
  HideModal,
  IntoGameDetail,
  OutOfGameDetail,
  PagedNavigation,
  ToggleOn,
  ToggleOff,
  SliderUp,
  SliderDown,
  ChangeTabs,
  DefaultOk,
  OpenSideMenu,
  CloseSideMenu,
  BasicNav,
  FailedNav,
  Typing,
  TimerExpired,
  Screenshot,
}
