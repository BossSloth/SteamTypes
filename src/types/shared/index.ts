export * from './enums';
export * from './interfaces';
export * from './protobuf';
export * from './steamid';
