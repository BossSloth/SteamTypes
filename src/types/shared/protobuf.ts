import { ExtensionFieldInfo } from 'google-protobuf';
import { CMsgProtoBufHeader } from '../Protobufs/steam/steammessages_base';

/**
 * To get the deserialized object call {@link ProtobufNotification.Body()}.toObject()
 * @todo: Find way to add this to the compare maps to check for new types
 */
export interface ProtobufNotification<T> {
  BIsValid(): boolean;

  Body(): ProtobufInterface<T>;

  BSuccess(): boolean;

  DEBUG_LogToConsole(): void;

  DEBUG_ToObject(): object | unknown;

  GetEMsg(): number;

  GetEResult(): number;

  GetErrorMessage(): string;

  Hdr(): Header;

  InitForType(e: unknown): void;

  ReadBodyFromBuffer(e: unknown, t: unknown): void;

  Serialize(): Uint8Array;

  SerializeBody(): Uint8Array;

  SetBodyFields(e: unknown): void;

  SetBodyJSON(e: unknown): void;

  SetEMsg(e: number): void;

  m_body: Body;

  m_bValid: boolean;

  m_cubHeader: number;

  /**
   * This value is an enum
   * @currentValue 146
   */
  m_eMsg: number;

  m_header: Header;

  m_netPacket: NetPacket;
}

export type Header = ProtobufInterface<CMsgProtoBufHeader>;

export interface NetPacket {
  GetCountBytesRemaining(): number;

  GetPacket(): unknown;

  GetUint8(): unknown;

  /**
   * @param e default: !0
   */
  GetUint32(e?: boolean): unknown;

  PutBytes(e: unknown): void;

  PutUint8(e: unknown): void;

  /**
   * @param t default: !0
   */
  PutUint32(e: unknown, t?: boolean): void;

  SeekGetCurrent(e: unknown): void;

  /**
   * @param e default: 0
   */
  SeekGetHead(e?: number): void;

  SeekPut(e: unknown): void;

  TellGet(): unknown;

  TellMaxPut(): unknown;

  TellPut(): unknown;

  m_iGet: number;

  m_iPut: number;

  m_nLength: number;

  m_nOffset: number;

  m_rgubPacket: Uint8Array;

  m_viewPacket: DataView;
}

type ProtoGetters<T> = {
  [K in keyof T]-?: () => Exclude<T[K], undefined>;
};

type ProtoSetters<T> = {
  [K in keyof T as `set_${string & K}`]-?: (value: T[K]) => ProtobufInterface<T>;
};

// Helper to check if a type is an interface (object but not array, primitive, or built-in)
type IsInterface<T> = T extends unknown[]
  ? false
  : T extends object
    ? T extends Function
      ? false
      : T extends Date | RegExp | Map<unknown, unknown> | Set<unknown> | WeakMap<object, unknown> | WeakSet<object> | Promise<unknown>
        ? false
        : true
    : false;

// Extract element type from array
type ArrayElement<T> = T extends (infer E)[] ? E : never;

// Conditional type for add method parameter - wraps interfaces in ProtobufInterface
type AddMethodParam<T> = IsInterface<T> extends true ? ProtobufInterface<T> : T;

// Add methods for all array properties
type ProtoAdders<T> = {
  [K in keyof T as Exclude<T[K], undefined> extends unknown[]
    ? `add_${string & K}`
    : never]: (
    value: AddMethodParam<ArrayElement<Exclude<T[K], undefined>>>,
  ) => void;
};

/**
 * Protobuf interface with getters, setters, and add methods for array properties
 * @note You probably don't want to use most of the properties here and
 * instead use the {@link ProtobufInterface.toObject()} method to convert the protobuf message to a plain object
 */
export type ProtobufInterface<T> = ProtoGetters<T> & ProtoSetters<T> & ProtoAdders<T> & SimpleJsPbMessage<T>;

export interface SimpleJsPbMessage<T> {
  clone(): this;

  cloneMessage(): this;

  /**
   * Returns the exact name of the protobuf message "class"
   */
  getClassName(): string;

  getExtension<T>(fieldInfo: ExtensionFieldInfo<T>): T;

  getJsPbMessageId(): unknown | undefined;

  serializeBase64String(): string;

  serializeBinary(): Uint8Array;

  setExtension<T>(fieldInfo: ExtensionFieldInfo<T>, value: T): void;

  toArray(): SimpleJsPbMessage<T>['array'];

  toObject(e?: boolean): T;

  array: InterfaceToTuple<T>;

  arrayIndexOffset_: number;

  /**
   * SimpleJsPbMessage[] of all nested objects in this message
   */
  wrappers_: SimpleJsPbMessage<unknown>[];
}

type InterfaceToTuple<T> = T extends object
  ? {
      [K in keyof T]-?: Exclude<T[K], undefined> extends infer V
        ? V extends object
          ? V extends unknown[]
            ? V // Keep arrays as-is
            : InterfaceToTuple<V> // Recursively convert nested interfaces
          : V
        : never;
    }[keyof T][]
  : T;
