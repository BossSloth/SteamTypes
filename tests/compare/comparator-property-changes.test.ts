import { createTest } from './shared';
import { propertyChangeCases } from './test-cases';

createTest('Property Change Cases', propertyChangeCases);
