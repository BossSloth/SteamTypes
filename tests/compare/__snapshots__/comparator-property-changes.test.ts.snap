// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Comparator Property Change Cases > flags enum with missing value 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -6,6 +6,8 @@

export enum EStatusFlags {
  CONNECTED = 1 << 0,
  DISCONNECTED = 1 << 1,
  PENDING = 1 << 2,
+  /** @generated bitfield value */
+  Flag_12 = 12
}"
`;

exports[`Comparator Property Change Cases > generic type parameter changes 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,25 +1,37 @@
import { ObservableMap } from 'mobx';

export interface DataStore {
-  cache: Set<string>;
+  cache: Set<number>;
+
  items: Map<string, Item>;
-  reactive: ObservableMap<string, ReactiveData>;
+
+  reactive: ObservableMap<number, ReactiveData>;
+
  users: ObservableSet<UserData>;
}

export interface Item {
+  data: string;
+
  id: string;
+
  name: string;
+
+  type: 'text' | 'image' | 'file';
}

export interface UserData {
+  avatarUrl: string;
+
  id: number;
+
  name: string;
-  email: string;
}

export interface ReactiveData {
  id: number;
+
  name: string;
+
  tracking: boolean;
}"
`;

exports[`Comparator Property Change Cases > generic type parameter changes with union types 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,29 +1,53 @@
import { ObservableMap } from 'mobx';

export interface DataCollection {
-  cache: ObservableMap<string, BasicItem | ExtendedItem>;
+  cache: ObservableMap<string, BasicItem | ExtendedItem | SpecialItem>;
+
  items: Set<ReactiveData | ExtendedItem>;
+
  mappings: Map<string, UserData | ReactiveData>;
}

export interface BasicItem {
+  createdAt: Date;
+
  id: string;
+
  name: string;
}

export interface ExtendedItem {
  description: string;
+
  id: string;
+
+  isActive: boolean;
+
  metadata: Record<string, unknown>;
+
  name: string;
}

export interface UserData {
+  avatarUrl: string;
+
  id: number;
+
  name: string;
}

export interface ReactiveData {
  id: number;
+
+  name: string;
+
  tracking: boolean;
}
+
+export interface SpecialItem {
+  id: string;
+
+  priority: number;
+
+  type: 'special';
+}"
`;

exports[`Comparator Property Change Cases > generic type with typeof string mismatch 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,5 +1,5 @@
const MAIN_WINDOW_NAME = 'SP Desktop_uid0';

export interface Container {
-  data: Map<typeof MAIN_WINDOW_NAME, string>;
+  data: Map<number, string>;
}"
`;

exports[`Comparator Property Change Cases > interface with added nested interfaces 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,7 +1,31 @@
export interface AppConfiguration {
+  features: FeatureFlags;
+
+  settings: AppSettings;
+
  version: string;
-  settings: {
-    theme: string;
-    language: string;
-  };
}
+
+export interface AppSettings {
+  language: string;
+
+  notifications: NotificationSettings;
+
+  theme: string;
+}
+
+export interface NotificationSettings {
+  desktop: boolean;
+
+  enabled: boolean;
+
+  sound: boolean;
+}
+
+export interface FeatureFlags {
+  betaFeatures: boolean;
+
+  darkMode: boolean;
+
+  experimentalApi: boolean;
+}"
`;

exports[`Comparator Property Change Cases > interface with implied enum property 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,4 +1,16 @@
export interface SteamStatus {
-  m_eResult: number;
  m_bRequireRestart: boolean;
+
+  /**
+   * This value is an enum
+   * @currentValue 0
+   */
+  m_eAppUpdateBytes: number;
+
+  m_eResult: EResult;
}
+
+/** @generated */
+export enum EResult {
+  EResult1 = 1
+}"
`;

exports[`Comparator Property Change Cases > interface with modified nested interfaces 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,12 +1,28 @@
export interface NetworkConfig {
  baseUrl: string;
+
  connection: ConnectionParams;
+
+  endpoints: Record<string, string>;
}

export interface ConnectionParams {
+  keepAlive: boolean;
+
+  proxy?: ProxySettings;
+
+  retries: number;
+
  /**
   * Timeout in milliseconds
   */
  timeout: number;
-  retries: number;
}
+
+export interface ProxySettings {
+  auth?: string;
+
+  host: string;
+
+  port: number;
+}"
`;

exports[`Comparator Property Change Cases > interface with property changing from one interface to another 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,11 +1,23 @@
export interface ChatMessage {
-  content: TextContent;
+  content: MessageContent;
+
  id: string;
+
+  read: boolean;
+
  sender: string;
+
  timestamp: number;
}

-export interface TextContent {
+export interface MessageContent {
+  data: TextData;
+
+  type: 'text' | 'image' | 'file';
+}
+
+export interface TextData {
+  format: 'plain' | 'markdown' | 'html';
+
  text: string;
-  contentFormat: 'plain' | 'markdown';
}"
`;

exports[`Comparator Property Change Cases > interface with property changing from required to optional 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,6 +1,13 @@
export interface ProductDetails {
-  description: string;
+  description?: string;
+
+  discount?: number;
+
  id: string;
+
+  inStock: boolean;
+
  name: string;
+
  price: number;
}"
`;

exports[`Comparator Property Change Cases > property changes from array to specific interface 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,9 +1,35 @@
export interface LibraryContent {
-  books: Book[];
-  magazines: any[];
+  books: BookCollection;
+
+  magazines: MagazineCollection;
}

export interface Book {
  title: string;
+
  author: string;
}
+
+export interface BookCollection {
+  count: number;
+
+  items: Book[];
+
+  lastUpdated: Date;
+}
+
+export interface MagazineCollection {
+  count: number;
+
+  items: Magazine[];
+
+  lastUpdated: Date;
+}
+
+export interface Magazine {
+  issue: number;
+
+  publisher: string;
+
+  title: string;
+}"
`;

exports[`Comparator Property Change Cases > property type changes from interface to primitive 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,10 +1,5 @@
export interface GameState {
-  player: PlayerStats;
+  player: string;
+
  gameMode: string;
}
-
-export interface PlayerStats {
-  score: number;
-  level: number;
-  achievements: string[];
-}"
`;

exports[`Comparator Property Change Cases > property type changes from primitive to interface 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,4 +1,13 @@
export interface UserSettings {
-  profile: string;
+  profile: UserProfile;
+
  preferences: Record<string, boolean>;
}
+
+export interface UserProfile {
+  avatar: string;
+
+  bio: string;
+
+  displayName: string;
+}"
`;

exports[`Comparator Property Change Cases > property with enum value changes 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,10 +1,14 @@
export interface StatusConfig {
-  status: ConnectionStatus;
+  errorCode?: number;
+
  message: string;
+
+  status: ConnectionStatus;
}

export enum ConnectionStatus {
  CONNECTED = 1,
  DISCONNECTED = 2,
  PENDING = 3,
+  "TODO: change me 5" = 5
}"
`;

exports[`Comparator Property Change Cases > property with shorter enum value changes 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,10 +1,14 @@
export interface StatusConfig {
+  errorCode?: number;
+
  eStatus: ConnectionStatus;
+
  message: string;
}

export enum ConnectionStatus {
  CONNECTED = 1,
  DISCONNECTED = 2,
  PENDING = 3,
+  "TODO: change me 5" = 5
}"
`;

exports[`Comparator Property Change Cases > shorted implied enum property 1`] = `
"===================================================================
--- a/target.ts
+++ b/target.ts
@@ -1,3 +1,14 @@
export interface SteamStatus {
-  eResult: number;
+  /**
+   * This value is an enum
+   * @currentValue 0
+   */
+  eAppUpdateBytes: number;
+
+  eResult: EResult;
}
+
+/** @generated */
+export enum EResult {
+  EResult1 = 1
+}"
`;
